Audit Logging Support on Sync Gateway (#803)
couchbaselabs/docs-sync-gateway: modules/ROOT/pages/manage-audit-logs.adoc

Reviewer: Still WIP?

Author: Still yes.

= Manage Audit Logs
:page-edition: Enterprise
:description: Administrators can manage audit logs to track operational irregularities and to support regulatory and security compliance standards.

[abstract]
{description}

> == Audit Fields 

Reviewer 2: This section is providing the definition for a subset of the properties that may be present on events.  I'm not sure this belongs in the 'Manage Audit Logs' section, as these properties aren't used to configure or manage audit logging.

> The table below lists mandatory audit fields with corresponding descriptions. 

Reviewer 2: I'm not sure about the use of the term 'mandatory' here.  Although these are common properties, it's not the case that every audit event emitted includes all of these properties.

Author: Changed!

Different event-types generate different field-subsets.

[cols="3*"]
|===

| Field | Type | Description

|`"id"`
|Integer
|The unique identifier of the event.

|`"cid"`
|Integer
|Context id for an individual request related events

|`"description"`
|String
|The name of the event in an action-based language: e.g. “User was created”.

|`"db"`
|String
|Name of database the event originated from.

|`"real_userid"`
|Object
|Contains key-value pairs for "domain" and "user" (specifying the id). Can be user specified or Server roles.

|`"local"`
|Object
|Contains key-value pairs for "ip" and incoming "port", for the node on which the event was processed.

|`"remote"`
|Object
|Contains key-value pairs for "ip" and outgoing "port", for the node on which the event-request was dispatched.

|`"timestamp"`
|ISO-8601 timestamp
|UTC timestamp for when the event was generated.

|`"optional fields"`
|Object
|Additional non-action or user id specific event data: e.g: {replication_type: pull_only}

|===

== Audit Log Configuration

IMPORTANT: Audit Logging on Sync Gateway is opt-in functionality and is disabled by default.

> Admin users can manage Audit logging via the xref:rest-api-admin.adocp[Admin REST API]. 

Reviewer 2: There are three main ways for users to manage Audit logging:
- bootstrap config (file-based config)
  - global configuration of audit logging
- database config (via existing Admin REST API endpoints)
   - per database configuration of audit logging as part of the database config
- via the new audit-specific Admin REST API endpoint (_config/audit)
    - per database configuration of audit logging only

I think it might make this section easier to follow if this was explained first, before going into the breakdown of each section.  The existing opening sentence ("users can manage Audit logging via the Admin REST API") makes it sound like this is the only way to do it.

Author: Changed

> Below is a table of possible configuration options.

Reviewer 2: It's not obvious which configuration option these properties are intended to apply to, or what they correspond to in the config file or REST API payloads.  I'd suggest removing this table and instead providing specific definitions for each of the three configuration options.

The properties in this table also don't align with the implemented product - I think switching to documenting the bootstrap config, database config and audit REST endpoint will avoid that problem

| Default - 1GB
| The maximum size the active audit logging file is allowed to grow before being rotated.

| `Rotation Interval`
| The time interval in seconds from 900s (15 minutes) to 604800s (7 days) inclusive. 
| The default is 86400s.

| `Prune age`
| The time after which Sync Gateway deletes audit rotates audit log files in minutes. 
| The default is 30 days (43200)

| `Max size`
| A specified filesize outside of the `Max File Size` property.
| This is disabled by default.

|===

There are two main ways in which you can configure audit logs:

* Via the bootstrap config.
This will affect the global configuration.

* Via the database config.
This will affect the configuration on a per database basis.

For more information on the full configuration schema see:

* xref:configuration-schema-database.adoc[Database Configuration Schema].

* xref:configuration-schema-bootstrap.adoc[Bootstrap Configuration Schema].

=== Configuration with the Bootstrap Config

Below you can find the minimal configuration required in your bootstrap config file to enable audit logging.

[source, json, indent=0]
----

"logging": {
    "audit": {
	  “audit_log_file_path”: “/path/to/file.log”,

>     “enabled”: true,

Reviewer 2: ```suggestion
           “enabled”: true
```

    }
}

----

For further bootstrap configuration options, see xref:configuration-schema-bootstrap.adoc[].

NOTE: Configuration via the bootstrap config will affect your configuration globally.

> To successfully configure audit logging, you must set a file path for your audit logs and set the `enabled` parameter to `true`.

Reviewer 2: If `audit_log_file_path` is not set, SGW will use the more general logging log file path (set in the bootstrap config)

Enabling and disabling audit logging is handled by the Admin REST API at a cluster and Sync Gateway database level.
Setting enabled to true without any specified enabled events will only enable events with `DefaultEnabled` set to `true`. 

> To view the list of audit events, see xref:audit-log-events.adoc[]

Reviewer 2: It doesn't mention here (or elsewhere?) that the audit log will be written to a file named sg_audit.log.

=== Configuration with the Database Config

> You can alter the configuration per Sync Gateway database by changing the database config.

Reviewer 2: I think it's worth clarifying the hierarchy here.  By default, the bootstrap audit configuration applies to all databases.  Specifying audit logging configuration in the database config allows users to override the global settings for a particular database.  Given that hierarchy, I wonder if it makes sense to go into the additional details about the bootstrap config first (explaining filtering, etc) before circling back to explaining how these details can be configured at the db level.

                                                "name": "user1",
                                                "domain": "sgw"
                                              }
                                            ]
                        },
  "disabled_roles": [
                      {
                        "name": "role1",
                        "domain": "sgw"
                      }
                    ]
               }              
}  
----

For further database configuration options, see xref:configuration-schema-database.adoc[].

IMPORTANT: If you enable audit logging in your bootstrap config it is enabled globally.
Conversely, it is important to note that if audit logging is enabled in your database config but *not* your bootstrap config, no audit logging will occur.

Couchbase recommends you only use configuration at the database level if you wish to alter your settings per database. 

== Filtering Audit Logs

You can filter audit logs to customize events that are stored in the log file. 
You can alter the following to affect events stored in the log file:

* Enabling and disabling users - This affects all events related to a specified user.

> If a given user is disabled, no audit events from the user will be logged.

Reviewer 2: ```suggestion
If a given user is disabled, no audit events associated with the user will be logged.
```
I feel like 'associated with' better describes the scenario than 'from' - 'from' sounds like the user is creating an audit event, when really the audit event is associated with some user action.

* Enabling and disabling events - This affects the specified events globally.
If a given event is disabled, it will not be logged regardless of the user.

> NOTE: If an audit event has `DefaultEnabled` set to `true`, it does not need to be explicitly listed in the `enabled_events: []` array.

Reviewer 2: I don't think this is correct.  If enabled_events is specified in the bootstrap config, that overrides the default set of enabled events.  Only if enabled_events is omitted will the default set of optional events be enabled.
https://confluence.issues.couchbase.com/wiki/spaces/SWD/pages/2387378237/SGW+3.2.0+Audit+Logging#Bootstrap%2Fstartup-config

See xref:audit-log-events.adoc[] for more information on specific audit events.

=== Enabled and Disabled Events

Events can be enabled per database or globally.

The example below shows a section of a global bootstrap config which would enable Public HTTP API requests and Public API user authenticated events only.

[source, json]
----
"logging": {
             "audit": {
                        “audit_log_file_path”: “/path/to/file.log”,
                        “enabled”: true,
                        “enabled_events”: [53270, 53280]
                      }
            }
----

> Enabling the events at runtime without modifying the entire database

Reviewer 2: At the end of the day these endpoints are going to modify the database (by updating the audit config for the database) - so this description isn't exactly right.  It would be better described as something like 'Modifying the database audit logging config independently'

> You can enable events at runtime without modifying the entire database using the following HTTP request methods:

Reviewer 2: Since 'GET' doesn't enable events, I might describe this as 'You can manage the audit logging configuration for a database....'

* `GET /{db}/_config/audit`
* `PUT /{db}/_config/audit (replace)`
* `POST /{db}/_config/audit (upsert)`

These request methods apply changes at the database level, a possible use case is that you can modify the `enabled` parameter to disable all events for the database.

The examples below show two ways to specify audit events.

The first example enables the `Public HTTP API request` and `Public API user authenticated` events.

[source, json]
----
{
	“enabled”: true,

>   “enabled_events”: [{“53270”: true, “53280”: true]

Reviewer 2: Name and format of the events property isn't quite right here, and should be indented to match 'enabled'.

```suggestion
        “events”: {“53270”: true, “53280”: true}
```

}
----

If you wanted to enable one event and disable another:

[source, json]
----
{
	“enabled”: true,
  “enabled_events”: [{“53270”: true, “53280”: false]
}
----

=== Disabled Users and Roles

You can filter audit events by specifying roles or users to be disabled. 
The example configuration below shows disabling audit events for `user1` and `role1`. 

[source, json]
----
{
  “logging”: {
               “audit”: {
                          "enabled”: true,

>                           “enabled_events”: [],

Reviewer 2: ```suggestion
                          “events”: {},
```

                                                “domain”: “sgw”
                                              }
                                            ],
                        },
  “disabled_roles”: [
                      {
                        “name”: “role1”,
                        “domain”: “sgw”
                      }
                    ],
               },
}  
----

The `disabled_users` field will prevent all audit events generated by the specified users from being logged.
The `disabled_roles` field will prevent all audit events generated by the specified roles from logged.
A use case for these fields would be to exclude certain administrative users or roles that perform a large volume of automated processes to prevent bloat of trivial events causing early rotation of the log file.

Users and roles are organised into the following domains:

* `sgw` - Users and Roles that are created by and operate solely within Sync Gateway.
For more information, see xref:access-control-concepts.adoc#lbl-sgw-users[Sync Gateway defined Users and Roles]. 
* `cbs` - Users that are are RBAC controlled.
These are created on Couchbase Server.

> For more information, see xref:access-control-concepts.adoc#lbl-rbac-users[RBAC Users]. 

Reviewer 2: It's probably worth explaining how the Couchbase Server users interact with Sync Gateway.  Sync Gateway's admin and metrics REST API are authenticated using Couchbase Server users.

= Manage Audit Logs
:page-edition: Enterprise
:description: Administrators can manage audit logs to track operational irregularities and to support regulatory and security compliance standards.

[abstract]
{description}

== Audit Log Configuration

IMPORTANT: Audit Logging on Sync Gateway is opt-in functionality and is disabled by default.

Admin users can configure audit logs using the following:

* Via the bootstrap config.
This will affect the global configuration.

* Via the Admin REST API.
This will affect the configuration on a per database basis.

For more information, see:

* xref:configuration-schema-database.adoc[Database Configuration Schema].

* xref:rest_api_admin.adoc[Admin REST API].

== Global Configuration with the Bootstrap Config

Below you can find the minimal configuration required in your bootstrap config file to enable audit logging.

[source, json, indent=0]
----

"logging": {
    "audit": {
	  "audit_log_file_path": "/path/to/file.log",

>     "enabled": true,

Reviewer 2: Needs alignment tweaked, "enabled" should be left-aligned with "audit_log_file_path".

    }
}

----

NOTE: Configuration via the bootstrap config will affect your configuration globally.

To successfully configure audit logging, you must set a file path for your audit logs and set the `enabled` parameter to `true`.
Enabling and disabling audit logging is handled by the Admin REST API at a cluster and Sync Gateway database level.
Setting enabled to true without any specified enabled events will only enable events with `DefaultEnabled` set to `true`. 
To view the list of audit events, see xref:audit-log-events.adoc[]

To configure audit logs retention and rotation globally, you can use the following parameters below with your bootstrap configuration.

[cols="3*"]
|===

> | Name | Values | Additional Description

Reviewer 2: This table is incorrect and doesn't represent the audit logging options present in the bootstrap configuration.  

Use Audit-logging-config (https://github.com/couchbase/sync_gateway/blob/main/docs/api/components/schemas.yaml#L2737) as the source of truth for bootstrap audit logging config format and properties.

Author: Fixed

= Manage Audit Logs
:page-edition: Enterprise
:description: Administrators can manage audit logs to track operational irregularities and to support regulatory and security compliance standards.

[abstract]
{description}

== Audit Log Configuration

IMPORTANT: Audit Logging on Sync Gateway is opt-in functionality and is disabled by default.

Admin users can configure audit logs using the following:

* Via the bootstrap config (file based config).
This will affect the global configuration of audit logging.

* Via the database config using existing Admin REST API endpoints.
This will affect per database configuration as part of the database config.

* Via the audit-specific Admin REST API endpoint `config/audit`.
This will affect the configuration on a per database basis.

For more information, see:

* xref:configuration-schema-bootstrap.adoc[Bootstrap Configuration Schema].

* xref:configuration-schema-database.adoc[Database Configuration Schema].

* xref:rest_api_admin.adoc[Admin REST API].

== Global Configuration with the Bootstrap Config

Below you can find the minimal configuration required in your bootstrap config file to enable audit logging.

[source, json, indent=0]
----

> "logging": {
>     "audit": {
>                 "audit_log_file_path": "/path/to/file.log",
>                 "enabled": true,
>              }
> }

Reviewer 2: ```suggestion
"logging": {
    "audit": {
        "audit_log_file_path": "/path/to/file.log",
        "enabled": true
    }
}
```

* xref:rest_api_admin.adoc#tag/Database-Configuration/operation/get_db-_config-audit[GET /{db}/_config/audit]
* xref:rest_api_admin.adoc#tag/Database-Configuration/operation/put_db-_config-audit[PUT /{db}/_config/audit (replace)]
* xref:rest_api_admin.adoc#tag/Database-Configuration/operation/post_db-_config-audit[POST /{db}/_config/audit (upsert)]

These request methods apply changes at the database level.
Below you can find examples demonstrating use of the above methods.

IMPORTANT: If you enable audit logging in your bootstrap config it is enabled globally.
Conversely, it is important to note that if audit logging is enabled in your database config but *not* your bootstrap config, no audit logging will occur.

.Get the database audit configuration
====
In this example:

* `$BASEURL` is the protocol and host for the Sync Gateway Admin API, for example `http://localhost`.
* `$DB` is the name of the database.

.Request
[source,shell]
----
curl "$BASEURL:4985/$DB/_config/audit"
----

.Response
[source,json]
----

> {
> 
>     "enabled": true,
>     "events": 
> {
> 
>     "53270": true,
>     "53271": true
> 
> },
> "disabled_users": 
> [
> 
>     {
>         "domain": "cbs",
>         "name": "user1"
>     }
> 
> ],
> "disabled_roles": 
> [
> 
>         {
>             "domain": "sgw",
>             "name": "role1"
>         }
>     ]
> 
> }

Reviewer 2: ```suggestion
{
    "enabled": true,
    "events": {
        "53270": true,
        "53271": true
    },
    "disabled_users": [
        {
            "domain": "cbs",
            "name": "user1"
        }
    ],
    "disabled_roles": [
        {
            "domain": "sgw",
            "name": "role1"
        }
    ]
}
```

Below is an example of a simple database configuration.

[source, json]
----
{
  "bucket": "todo", 
  "num_index_replicas": 0,
  “logging”: {
               “audit”: {
                          "enabled”: false,
                          “enabled_events”: [],
                          “disabled_users”: [
                                              {
                                                “name”: “user1”,
                                                “domain”: “sgw”
                                              }
                                            ],
                        },
  “disabled_roles”: [
                      {
                        “name”: “role1”,
                        “domain”: “sgw”
                      }
                    ],

>                },

Reviewer 3: This json isn't valid, can you pop this into a validator?

1. there is a trailing comma on the last element
2. there are smart quotes on the keys
3. there's missing a `}` to close `"logging"`

Author: Done. Forgot how fiddly JSON validation was!
